### PR TITLE
Create JWT [WIP]

### PR DESCRIPTION
**Pull request by aldigjo**
- [x] Add minimal signer interface
- [x] Add createJWT function
- [x] Add tests to match Android
- [x] flesh out signer interface and implementation class
- [x] Refactor payload to use NSMutableDictionary

### CODE REVIEW

File: UPort/JWT/JWTTools.swift [hunk 1]
```diff
@@ -76,6 +76,60 @@ public struct JWTTools
         }
     }
 
+    public static func create(payload: [String: Any],
```
Comment (pelle) [inline UPort/JWT/JWTTools.swift:97]: Shouldn't the payload be a Dict?

File: UPort/JWT/JWTTools.swift [hunk 2]
```diff
@@ -76,6 +76,60 @@ public struct JWTTools
...
+                {
+                    let rData = try? (sig!["r"] as? String)?.decodeBase64()
+                    let sData = try? (sig!["s"] as? String)?.decodeBase64()
+                    var vNum = sig!["v"] as? Int
+                    let vData = UnsafeBufferPointer(start: &vNum, count: 1)
+                    var rsv = Data()
+                    rsv.append(rData!!)
+                    rsv.append(sData!!)
+                    rsv.append(vData)
+                    let sigBase64 = rsv.base64EncodedString().replacingOccurrences(of: "=", with: "")
```
Comment (pelle) [inline UPort/JWT/JWTTools.swift]: How about moving `.replacingOccurrences(of: "=", with: "")` to `JWTTools.base64ToBase64Url()`

**Review by pelle — COMMENTED** [review]
A couple of minor questions, but otherwise looking good.

**Review by mirceanis — APPROVED** [review]
Looks good

File: UPort/JWT/JWTTools.swift [hunk 3]
```diff
@@ -76,6 +76,60 @@ public struct JWTTools
...
+        if filledOutPayload["exp"] == nil { filledOutPayload["exp"] = Int64(now().timeIntervalSince1970) + expiresIn }
+        filledOutPayload["iss"] = issuerDID
+        // Convert filled out payload to base64
+        let payloadBase64 = JWTTools.dictionaryToBase64(dict: filledOutPayload)
+        let payloadBase64Url = JWTTools.base64ToBase64Url(base64String: payloadBase64)
+
+        // Join the header/payload and base64 encode for signing
+        let signingInput = [headerBase64Url, payloadBase64Url].joined(separator: ".")
+        let signingData = signingInput.data(using: .utf8)
+        let signingInputBase64 = signingData?.base64EncodedString()//.replacingOccurrences(of: "=", with: "")
```
Comment (mirceanis) [inline UPort/JWT/JWTTools.swift]: comment at end of line should be cleaned

File: UPort/JWT/JWTTools.swift (hunk 1 above)
Comment (mirceanis) [inline UPort/JWT/JWTTools.swift:97]: This public method deserves a swiftdoc comment detailing the use and params.

Comment (mirceanis) [reply UPort/JWT/JWTTools.swift:97]: I think it already is

File: UPort/JWT/JWTTools.swift (hunk 2 above)
Comment (mirceanis) [reply UPort/JWT/JWTTools.swift]: sounds like a good idea

File: UPort/Signer/KPSigner.swift [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+//  UPort
+//
+//  Created by Aldi Gjoka on 4/4/19.
+//  Copyright © 2019 ConsenSys. All rights reserved.
+//
+
+import Foundation
+import UPTEthereumSigner
+
+class KPSigner: Signer {
```
Comment (mirceanis) [inline UPort/Signer/KPSigner.swift:13]: It is worth providing a swiftdoc comment for this class

File: UPort/Signer/Signer.swift [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+//
+//  Created by Aldi Gjoka on 3/7/19.
+//  Copyright © 2019 ConsenSys. All rights reserved.
+//
+
+import Foundation
+
+typealias SignatureCallback = (exception?, NSDictionary?) -> Void
+
+public protocol Signer {
```
Comment (mirceanis) [inline UPort/Signer/Signer.swift:14]: It's worth providing some swiftdoc comments for this protocol

**Review by mirceanis — CHANGES_REQUESTED** [review]
This looks good overall.
I am requesting some changes regarding more documentation for public methods and classes.

Also, it would be nice to have a test that creates a JWT and matches it against an expected hardcoded output.
This would be possible if the time were fixed for testing.

**Review by mirceanis — APPROVED** [review]
looks good

**Pull request merged**